src/include/hip/rpp_hip_common.hpp

```diff
@@ -72,4 +132,488 @@ inline RppStatus generate_gaussian_kernel_gpu(Rpp32f stdDev, Rpp32f* kernel, Rpp
 
     return RPP_SUCCESS;
 }
+
+/******************** DEVICE FUNCTIONS ********************/
+
+// -------------------- Set 1 - Packing --------------------
+
+// Packing to U8s
+
+__device__ __forceinline__ uint rpp_hip_pack(float4 src)
+{
+    return __builtin_amdgcn_cvt_pk_u8_f32(src.w, 3,
+           __builtin_amdgcn_cvt_pk_u8_f32(src.z, 2,
+           __builtin_amdgcn_cvt_pk_u8_f32(src.y, 1,
+           __builtin_amdgcn_cvt_pk_u8_f32(src.x, 0, 0))));
+}
+
+// Packing to I8s
+
+__device__ __forceinline__ uint rpp_hip_pack_i8(float4 src)
+{
+    char4 dst_c4;
+    dst_c4.w = (signed char)(src.w);
+    dst_c4.z = (signed char)(src.z);
+    dst_c4.y = (signed char)(src.y);
+    dst_c4.x = (signed char)(src.x);
+
+    return *(uint *)&dst_c4;
+}
+
+// -------------------- Set 2 - Un-Packing --------------------
+
+// Un-Packing from U8s
+
+__device__ __forceinline__ float rpp_hip_unpack0(uint src)
+{
+    return (float)(src & 0xFF);
+}
+
+__device__ __forceinline__ float rpp_hip_unpack1(uint src)
+{
+    return (float)((src >> 8) & 0xFF);
+}
+
+__device__ __forceinline__ float rpp_hip_unpack2(uint src)
+{
+    return (float)((src >> 16) & 0xFF);
+}
+
+__device__ __forceinline__ float rpp_hip_unpack3(uint src)
+{
+    return (float)((src >> 24) & 0xFF);
+}
+
+__device__ __forceinline__ float4 rpp_hip_unpack(uint src)
+{
+    return make_float4(rpp_hip_unpack0(src), rpp_hip_unpack1(src), rpp_hip_unpack2(src), rpp_hip_unpack3(src));
+}
+
+// Un-Packing from I8s
+
+__device__ __forceinline__ float rpp_hip_unpack0(int src)
+{
+    return (float)(signed char)(src & 0xFF);
```

Comment:
why the casting to (signed char) is needed here?

Reply:
Actually I found this to be the only way to do the unpack correctly. It takes  src, gets the first byte by masking with 0xFF, then we explicitly tell the compiler to interpret the MSB as a signed bit, then convert to float. If I directly convert to float, the most significant bit was interpreted not as a sign, and as part of an 8 bit number.